### ISSUE TITLE
Transforming ITK-style warping field to FSL/SPM-style produces error when used

### ISSUE THREAD

**Issue by jcohenadad** [neuropoly/spinalcordtoolbox#2525]
When trying to convert the ITK-based warping field (produced by SCT) to an FSL/SPM-style, as instructed in the [SCT course (page 49), version:5](https://osf.io/arfv7/?version=5) it produces an error when trying to use it. Below the code to reproduce the error (based on `batch_processing.sh` output):

Tested on OSX Mojave with SCT 4.1.0.

Syntax:
~~~
cd sct_example_data/t2
sct_image -i warp_template2anat.nii.gz -mcs -o warp_vec.nii.gz
sct_image -i warp_vec_*.nii.gz -concat t -o warp_template2anat_fsl.nii.gz 
applywarp -i /Users/julien/code/sct/data/PAM50/template/PAM50_t2.nii.gz -r t2.nii.gz -w warp_template2anat_fsl.nii.gz -o PAM50_t2_reg_fsl.nii.gz
~~~

Error:
~~~
error: inv(): matrix appears to be singular

An error occured while reading file: warp_template2anat_fsl.nii.gz
~~~

Related to: http://forum.spinalcordmri.org/t/errors-converting-sct-generated-warps-for-use-with-fsl/233

**Comment by jcohenadad** [comment]
Digging a bit further, i've noticed that the order of the warping is inverted between SCT and FSL, i.e., to apply get A->B, we need to apply warp_B2A (not warp_A2B as in SCT).

Also, I tried running applywarp with the option `--rel`, and it does produce "better" results, although there is still an affine transformation missing (or not being properly applied):
~~~
applywarp -r t2.nii.gz -i t2_straight.nii.gz -w warp_curve2straight_fsl.nii.gz -o t2_straight_fsl.nii.gz --rel
~~~

It might be because of this reversed order: the affine matrix and the warping field should be re-concatenated but in the reverse order. 

To confirm this, I will generate a single warping field (without affine matrix), and see if I can replicate FSL and SCT warping.

**Comment by jcohenadad** [comment]
OK, it seems like the warping field is inverted in one dimension. By running the code below:
~~~
sct_straighten_spinalcord -i t2.nii.gz -s t2_seg.nii.gz
sct_apply_transfo -i t2.nii.gz -d straight_ref.nii.gz -w warp_curve2straight.nii.gz -o t2_straight.nii.gz
sct_image -i warp_straight2curve.nii.gz -mcs -o warp_straight2curve_vec.nii.gz
sct_image -i warp_straight2curve_vec_*.nii.gz -concat t -o warp_straight2curve_fsl.nii.gz 
applywarp -i t2_straight.nii.gz -r t2.nii.gz -w warp_straight2curve_fsl.nii.gz -o t2_straight_curved.nii.gz
~~~

The expected result is to get the straightened spinal cord warped back onto the curved t2 space. However, what is happening is that the straight cord is warped in the other direction instead:

original t2:
![Screen Shot 2019-11-14 at 2 51 22 PM](https://user-images.githubusercontent.com/2482071/68891125-405c5f00-06ee-11ea-94a2-96c32cef8ee2.png)

straight cord:
![Screen Shot 2019-11-14 at 2 51 39 PM](https://user-images.githubusercontent.com/2482071/68891143-48b49a00-06ee-11ea-8735-0cc7aa0d4c58.png)

straight cord warped on the t2 (wrong direction):
![Screen Shot 2019-11-14 at 2 51 52 PM](https://user-images.githubusercontent.com/2482071/68891188-5c600080-06ee-11ea-9c57-f535ed566312.png)

I suspect this is due to [this direction swap along the SI direction](https://github.com/neuropoly/spinalcordtoolbox/blob/master/spinalcordtoolbox/straightening.py#L461), which was done to fit ITK's requirement.

**Comment by jcohenadad** [comment]
I've tried to multiple the *_vec_Y file by -1, and even though the orientation of the cord makes more sense, there is still an undesired translation happening:
![Screen Shot 2019-11-14 at 3 09 02 PM](https://user-images.githubusercontent.com/2482071/68892941-16a53700-06f2-11ea-8b93-6eb49c95bf82.png)


At the end, making warping fields compatible between ITK and FSL might not be an easy task, as noted in [this post](https://sourceforge.net/p/advants/discussion/840261/thread/797583c5/#2f69/5b0d/ef81).

**Comment by jcohenadad** [comment]
Other references:
- https://www.jiscmail.ac.uk/cgi-bin/webadmin?A2=ind1509&L=FSL&D=0&P=178218
- https://www.mail-archive.com/hcp-users@humanconnectome.org/msg05794.html
- https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/FNIRT/UserGuide

**Comment by jcohenadad** [comment]
Digging further, I've made some progress when trying to generate an FSL-compatible warping field, when the src and dest space are the same. The code for testing is the following:
```bash
# Tested using SCT 4.1.0 and FSL 6.0.1
# Download testing data
sct_download_data -d sct_testing_data
cd sct_testing_data/t2
# Create labels 1 pixel apart for shifting the same image by one pixel in the X direction (in the image coordinate system)
sct_label_utils -i t2.nii.gz -create 30,27,25,1 -o label_src.nii
sct_label_utils -i t2.nii.gz -create 31,27,25,1 -o label_dest.nii
# Shift the image
sct_register_multimodal -i t2.nii.gz -ilabel label_src.nii -d t2.nii.gz -dlabel label_dest.nii -param step=0,type=label,dof=Tx_Ty:step=1,type=im,algo=syn,iter=0
# Register using ANTs bsplynesyn
sct_register_multimodal -i t2.nii.gz -d t2_src_reg.nii.gz -param step=1,type=im,algo=bsplinesyn,iter=5
# Convert ITK warp into FSL warp
sct_image -i warp_t2_src_reg2t2.nii.gz -mcs -o warp_t2_src_reg2t2_vec.nii.gz
sct_image -i warp_t2_src_reg2t2_vec*.nii.gz -concat t -o warp_t2_src_reg2t2_fsl.nii.gz 
# Apply transformation using FSL
applywarp -i t2_src_reg.nii.gz -r t2.nii.gz -o t2_src_reg_reg-fsl.nii.gz -w warp_t2_src_reg2t2_fsl.nii.gz
```

The results show that the applied warp moves the image along X, which is consistent with [previous reports](https://www.mail-archive.com/hcp-users@humanconnectome.org/msg05794.html) mentioning that ITK warps are defined in the world coordinate while FSL ones are in the image coordinate systems. In this case, image orientation is AIL, so X (in image space) corresponds to Y (in world space).

**Comment by jcohenadad** [comment]
OK I figured that out, the following code is working. It requires a modified version of SCT (to convert warping fields from world to vox coordinate system). Another notable thing is that FSL takes ITK's inverse warping field:
```bash
# Tested using SCT (jca/fsl-warp-7870ef216d3a67eaf547fd66ece3a0c5a687e4ee) and FSL 6.0.1
# Download testing data
sct_download_data -d sct_testing_data
cd sct_testing_data/t2
# Create labels 1 pixel apart for shifting the same image by one pixel in the X direction (in the image coordinate system)
sct_label_utils -i t2.nii.gz -create 30,27,25,1 -o label_src.nii
sct_label_utils -i t2.nii.gz -create 31,27,25,1 -o label_dest.nii
# Shift the image
sct_register_multimodal -i t2.nii.gz -ilabel label_src.nii -d t2.nii.gz -dlabel label_dest.nii -param step=0,type=label,dof=Tx_Ty:step=1,type=im,algo=syn,iter=0
# Register using ANTs bsplynesyn
sct_register_multimodal -i t2.nii.gz -d t2_src_reg.nii.gz -param step=1,type=im,algo=bsplinesyn,iter=5
# Convert ITK warp into FSL warp
# Note: here we take the inverse warp because FSL treats it as the forward warp
sct_image -i warp_t22t2_src_reg.nii.gz -mcs -o warp_t22t2_src_reg_vec.nii.gz
sct_image -i warp_t22t2_src_reg_vec*.nii.gz -concat t -o warp_t22t2_src_reg_fsl.nii.gz 
# Convert displacement field to voxel (not world) coordinate system
sct_image -i warp_t22t2_src_reg_fsl.nii.gz -world2vox -o warp_t22t2_src_reg_fsl_vox.nii.gz
# Apply transformation using FSL
applywarp -i t2_src_reg.nii.gz -r t2.nii.gz -o t2_src_reg_reg-fsl2.nii.gz -w warp_t22t2_src_reg_fsl_vox.nii.gz
```

Next step is to check wether this code works for all situations (native image orientation, displacement direction, etc.)

**Comment by jcohenadad** [comment]
There are inconsistent results, depending if the determinant of the qform is positive (preserves left-right orientation) or negative (reverses left-right orientation). [This link](https://nifti.nimh.nih.gov/nifti-1/support/FSLandNIfTI1/) explains that regardless of the determinant, FSL treats nifti files with reversed left-right orientation, so this should be accounted for.

To reproduce, use the code above and simply change the orientation of the input file:
~~~
sct_image -i t2.nii.gz -setorient RPI -o t2.nii.gz  # will work
sct_image -i t2.nii.gz -setorient LPI -o t2.nii.gz  # will not work
~~~

**Comment by jcohenadad** [comment]
Above issue fixed in 0c52621a7f33fbb0d3bb6211472bff3f58c09fd9

**Comment by jcohenadad** [comment]
Now trying to make it work when src and dest space are different. The idea is to extract the affine transformation matrix, make it FSL compatible, concatenate the affine and the warp using FSL, then apply the final warp. 

Started some investigations, but currently not working:
~~~
# Tested using SCT (jca/fsl-warp-b066b489ddf7a5e798ce939ddb3fd05befb8994e) and FSL 6.0.1
# Straighen spinal cord
sct_straighten_spinalcord -i t2.nii.gz -s t2_seg.nii.gz
# Convert ITK warp into FSL warp
# Note: here we take the inverse warp because FSL treats it as the forward warp
sct_image -i warp_straight2curve.nii.gz -mcs -o warp_straight2curve_vec.nii.gz
sct_image -i warp_straight2curve_vec*.nii.gz -concat t -o warp_straight2curve_fsl.nii.gz 
# Convert displacement field to voxel (not world) coordinate system
sct_image -i warp_straight2curve_fsl.nii.gz -world2vox t2.nii.gz -o warp_straight2curve_fsl_vox.nii.gz
# Concatenate affine matrix and warping field
convertwarp -r t2.nii.gz -m affine_fsl.mat -w warp_straight2curve_fsl_vox.nii.gz -o warp_straight2curve_fsl_vox_withAffine.nii.gz
# Apply transformation using FSL
applywarp -i t2.nii.gz -r t2_straight.nii.gz -o t2_straight-fsl.nii.gz -w warp_straight2curve_fsl_vox_withAffine.nii.gz
~~~

Note: I've also tried those things (unsuccessful):
- input the matrix as a postmat (instead of premat), but doesn't work.
- use the inverse matrix (i.e. switched m and m_source [here](https://github.com/neuropoly/spinalcordtoolbox/blob/jca/fsl-warp/scripts/sct_image.py#L305))

**Comment by cfhammill** [comment]
So I've been trying to figure out how to fix this and I haven't found the answer yet. I tried a brute force approach to see if I could find a working transform and work backward from there but with no luck.

Surprising to me, the affine space of the inverse transform differs from the spaces of both the source and the target images. My intuition was that the inverse transform should be in the same space as the source image. 

I did a brute force search over potential affine registrations between these 3 spaces

- src->tgt
- tgt->src
- src->xfm
- xfm->src
- tgt->xfm
- xfm->tgt

these were each computed by extracting the affine matrices and solving for the transformation e.g.

```python
src = Image("t2.nii.gz")
tgt = Image("t2_straight.nii.gz")

src2tgt = np.linalg.solve(src.header.get_best_affine(), tgt.header.get_best_affine())
np.savetxt("src2tgt.mat", src2tgt)
```

I tried all 6 transforms with both `--premat` and `--postmat` in fsl `convertwarp` followed by `applywarp`.  e.g.

```bash
file=src2tgt.mat
convertwarp --warp1=test_xfm.nii.gz --ref=t2_straight.nii.gz --out=test_post_${file%%.mat} --postmat=$file
applywarp -r t2_straight.nii.gz -w test_post_${file%%.mat} -o t2_post_${file%%.mat} -i t2.nii.gz
```

None worked as well as just applying the transform as computed by `sct_image --world2vox`, although applying that transform had a small but noticeable rotation.

Any ideas on why the inverse transform isn't in the same space as the target? 

One alternative idea I saw on the [MRtrix forums](https://community.mrtrix.org/t/warping-tck-files-using-ants-warps/162/2) is to generate a world coordinate volume in the target space, transform it to the source space along the inverse transform. From there it's trivial to compute the displacement vectors voxelwise. Do you think that's worth a shot?

**Comment by jcohenadad** [comment]
Thank you for the investigations @cfhammill!

> Any ideas on why the inverse transform isn't in the same space as the target?

Based on the 2nd paragraph of your post above, your intuition was that the inverse transform would be defined in the coordinates of the source (not the target) image, right?  Regardless, it is indeed weird that the inverse transform is neither in the src nor target space. Also, did you mean that for  FSL or for ANTs? I'm actually not sure if this convention applies to both software (I know it is the case for ANTs, but is it also expected that the inverse transform is in the same space as the src image in FSL?). 

> One alternative idea I saw on the MRtrix forums is to generate a world coordinate volume in the target space, transform it to the source space along the inverse transform. From there it's trivial to compute the displacement vectors voxelwise. Do you think that's worth a shot?

Interesting post. Yes, at this point I would say that every direction is worth taking. If we can find a working solution, we might be able to reverse-engineering the procedure, understand it and simplify it. Also, don't hesitate to involve Mark Jenkinson from the FSL team.

**Comment by cfhammill** [comment]
Sorry yes, I meant the source image. This is the case for the ANTs transforms, to reproduce:

```python
from spinalcordtoolbox.image import Image

src = Image("t2.nii.gz")
tgt = Image("t2_straight.nii.gz")
s2c = Image("warp_straight2curve.nii.gz")
```

```python
print(s2c.header.get_best_affine() - src.header.get_best_affine())
```

```
[[ 0.          0.          0.          0.08375549]
 [ 0.          0.          0.          0.        ]
 [ 0.          0.          0.         -3.19890594]
 [ 0.          0.          0.          0.        ]]
```

```python
print(s2c.header.get_best_affine() - tgt.header.get_best_affine())
```

```
[[ 0.00000000e+00 -3.41523684e-18 -1.86264515e-09  0.00000000e+00]
 [-3.90429249e-21  0.00000000e+00  1.30971622e-16 -9.36000023e+01]
 [-1.86264515e-09  1.30537942e-16  0.00000000e+00  0.00000000e+00]
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]]
```

> I'm actually not sure if this convention applies to both software (I know it is the case for ANTs, but is it also expected that the inverse transform is in the same space as the src image in FSL?).

Not sure about the FSL convention, but if the convention was opposite for FSL I would expect a source to target affine transform to fix it (not sure if applied pre or post warping).

> Interesting post. Yes, at this point I would say that every direction is worth taking. If we can find a working solution, we might be able to reverse-engineering the procedure, understand it and simplify it. Also, don't hesitate to involve Mark Jenkinson from the FSL team.

Ok, I'll give it a shot this week and if I'm still unsuccessful I'll reach out to the FSL team.

**Comment by jcohenadad** [comment]
Another thing to keep in mind: some of SCT's transforms, including straightening, are not based on ANTs but are custom-made. It is possible that there are some bugs. So, if we are dealing with the ITK/FSL compatibility, I would start off with warping field that we trust better, e.g., those which are generated by ANTs. So, for those first investigations, I would instead work on warping fields that are generated by `sct_register_multimodal`, using ANTs-based transforms (e.g. SyN).

**Comment by cfhammill** [comment]
So, using `sct_register_multimodal` produces registrations that are in the same space as the source images (coordinate-wise difference < 10^-8).  The difference between the registration produced by `sct_register_multimodal` and `sct_straighten_spinalcord` is due to the padding operation in `sct_straighten_spinalcord`, which helps explain the small shift I was seeing when using `--world2vox`.

I think the answer is close.

**Comment by cfhammill** [comment]
Still haven't got it. Turns out that at least for `sct_register_multimodal` the transforms are stored as RAS coordinate relative transforms. So to determine where in the src image to grab intensity levels you do compute:

1. covert target voxel coordinate to world
2. add deformation vector (flipping sign for dimensions 1 and 2)
3. convert world coordinate to src voxel coordinate
4. interpolate

So I can do all of those steps in python to get pretty close to the right image for both the `sct_straighten_spinalcord` and `sct_register_multimodal`, so I have high confidence that my spatial transforms are correct. They match both the results from the initial registration, and subsequent resampling with `isct_antsApplyTransforms`. So I think the deformations produced by `sct_straighten_spinalcord` are trustworthy.

Now when using `applywarp`

I can write out the warp fields after converting to RPI and apply them with `applywarp`, provided I use `--premat=tgt2src.mat` [note that the tgt2src computed above is incorrect, this calculated by `np.linalg.solve(aff(src).T, aff(tgt).T).T` ].
 
The use of the `--premat` here doesn't make any sense to me, but produces a correct looking result.  Although it is possible the silly deformation field produced by:

```
sct_register_multimodal -i t2.nii.gz -d t2_straight.nii.gz \
  -o t2_straight_ants.nii.gz \
  -owarp ants_straight_to_curv.nii.gz -r 0
```

is too trivial to detect issues [ it doesn't look all that different from simply applying src2tgt.mat ].

However, when I use the exact same approach on the registrations produced by `sct_straighten_spinal_cord` it fails miserably.

Further, supplying the exact src coordinates for the deformation and using `applywarp --abs` doesn't work. So I'm unsure how the fsl deformations are interpreted. I skimmed the `fnirt` and `applywarp` code and nothing jumped out at me. Looks like I will probably need to reach out to Mark this week.
